File: store/115.不同的子序列.c
/*
 * @lc app=leetcode.cn id=115 lang=c
 *
 * [115] 不同的子序列
 */

// @lc code=start
int numDistinct(char *s, char *t)
{
  int i, j;
  unsigned int **dp;
  dp = (unsigned int **)malloc(sizeof(unsigned int *) * (strlen(s) + 1));
  for (i = 0; i < strlen(s) + 1; i++)
    dp[i] = (unsigned int *)malloc(sizeof(unsigned int) * (strlen(t) + 1));

  for (i = 0; i < strlen(s) + 1; i++)
    dp[i][0] = 1;
  for (j = 1; j < strlen(t) + 1; j++)
    dp[0][j] = 0;

  for (i = 1; i < strlen(s) + 1; i++)
  {
    for (j = 1; j < strlen(t) + 1; j++)
    {

      if (s[i - 1] == t[j - 1])
        dp[i][j] = dp[i - 1][j - 1] + dp[i - 1][j];
      else
        dp[i][j] = dp[i - 1][j];
    }
  }

  return dp[strlen(s)][strlen(t)];
}
// @lc code=end

File: store/115.不同的子序列.js
/*
 * @lc app=leetcode.cn id=115 lang=javascript
 *
 * [115] 不同的子序列
 */

// @lc code=start
/**
 * @param {string} s
 * @param {string} t
 * @return {number}
 */
var numDistinct = function (s, t) {
  let dp = new Array(s.length + 1)
    .fill()
    .map(() => Array(t.length + 1).fill(0));

  for (let i = 1; i < s.length + 1; i++) {
    if (s[i - 1] === t[0]) dp[i - 1][0] = 1;
  }
  dp[0][0] = 1;

  for (let i = 1; i < s.length + 1; i++) {
    for (let j = 1; j < t.length + 1; j++) {
      if (s[i - 1] === t[j - 1]) dp[i][j] = dp[i - 1][j - 1] + dp[i - 1][j];
      else dp[i][j] = dp[i - 1][j];
    }
  }

  return dp[s.length][t.length];
};
// @lc code=end
